### ISSUE TITLE
Improve parser to skip `#!` inside strings

### ISSUE THREAD

**Issue by fingolfin** [gap-packages/AutoDoc#126]
... so that we can get rid of `@DONT_SCAN_NEXT_LINE`.

Of course that's easier said than done. For "simple" strings, we can add a heuristic, say by counting the number of `"` we see on a line. But this doesn't work in all cases, e.g.:
```gap
x := "\"bla: #!";
y := " ab \
#!cd";
z := """Triple quotes are cool!
#! aren't they?
""";
```
The only way to overcome all that would be write a proper parser... I think @markuspf has done something in this direction, so perhaps we can collaborate with him on that in the futur?

In the meantime, though, we could just add a very simple heuristic: If `#!` is immediately preceded by a `"`, then do not treat it as a documentation comment. Of course that will annoy everytbody who for some strange reasons wants to write this:
```
#"#! @Section BLA
```
But, well, I don't feel sympathy ;-).

Of course we could also insist that AutoDoc comments must start at the beginning of the, and match a pattern like this: `^\s*#+! `. It could still be fooled by a multiline string, but that's no worse than what we have right now.

**Comment by fingolfin** [comment]
The last idea assumes that all doc comments *precede* whatever they are documenting. In Doxygen, you can also have comments on the same line, like so:
```C++
class Foo {
  int var; ///<  this is an important member variable
};
```
But this is a special syntax, so if we ever wanted to add something like this, we could also require a special syntax, e.g. `##<` or whatever.

**Comment by sebasguts** [comment]
I agree that the AutoDoc parser should be rewritten sometime, and then include some syntax awareness to GAP, e.g., recognizing if the AutoDoc tag is inside a string.

On the other hand, we might loose the possibility to include it in C++-Code, but I am not sure about this. Is @markuspf writing a parser for AutoDoc?

**Comment by fingolfin** [comment]
He has a parser for *GAP*. So that should be able to accurately tell what is a comment, and what is text inside a string. I don't think it knows about AutoDoc right now, but assuming @markuspf wrote it in a sane fashion (which I assume :-), it shouldn't be too hard to teach it about that.

**Comment by fingolfin** [comment]
If one has a tokenized version of the file, or even an AST, then also extracting things like filter arguments etc. becomes easier and less error prone.

**Comment by sebasguts** [comment]
Sure, if we would have the AST, we could very easy read out all the information out of it. I assume the parser is not written in GAP?
